rsciio/pantarhei/restricted_unpickling.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+# Copyright 2025 CEOS GmbH
+# Copyright 2022-2025 The HyperSpy developers
+#
+# This file is part of RosettaSciIO.
+#
+# RosettaSciIO is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# RosettaSciIO is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with RosettaSciIO. If not, see <https://www.gnu.org/licenses/>.
+"""Implements an unpickler, which restricts the allowed modules which pickle can load."""
+
+import pickle
+from logging import getLogger
+
+import numpy
+from packaging.version import Version
+
+numpy_version = Version(numpy.__version__)
+
+if numpy_version < Version("2.3"):
+    # for numpy < 2.3
+    from numpy.lib.format import _check_version, _read_array_header, read_magic
+else:
+    # for numpy >= 2.3
+    from numpy.lib._format_impl import _check_version, _read_array_header, read_magic
```

Comment:
Can we avoid importing private numpy API? There is some functionality that may be relevant in https://github.com/hyperspy/rosettasciio/blob/4d5db8a8c7b7e3909b6d7eb9951dfbcd1fdb47fe/rsciio/numpy/_api.py#L36C1-L37C1

Reply:
I think using the existing code you pointed to at https://github.com/hyperspy/rosettasciio/blob/4d5db8a8c7b7e3909b6d7eb9951dfbcd1fdb47fe/rsciio/numpy/_api.py#L36C1-L37C1 is the best idea to avoid private numpy API calls. I see two options:
1. Slight refactoring of `inspect_npy_file` to also accept byte streams (i.e from an npz) and import into `pantarhei._api` module. Probably best to move it to `utils` to not make the pantarhei api dependent on the numpy api implementation.
2. Copying of the relevant parts to a new function like `inspect_npy_bytes` residing directly in the`restricted_unpickling`  module.

What would you suggest?